decimal.go

```diff
@@ -301,6 +303,44 @@ func (d Decimal) Mod(d2 Decimal) Decimal {
 	return d.Sub(d2.Mul(quo))
 }
 
+// Sqrt returns the square root of d. The method will panic if d
+// is a negative number.
+//
+// The square root calculation is implemented using Newton's Method.
+// We start with an initial estimate for sqrt(d), and then iterate:
+//     x_{n+1} = 1/2 * ( x_n + (d / x_n) ).
+func (d Decimal) Sqrt() Decimal {
+	d.ensureInitialized()
+
+	// Validate the sign of d
+	switch d.BigInt().Sign() {
+	case -1:
+		panic(fmt.Sprintf("square root of negative number: %v", d))
+	case 0:
+		return Zero
+	}
+
+	// Use half as the initial estimate.
+	x := d.Mul(half)
+
+	// Determine the maximum number of iterations needed.
+	scale := len(d.BigInt().String())
```

Comment:
Can you derive this value from `d.BigInt().BitLen()` instead? Converting the bigint to a string seems like a waste.